src/server.zig

```diff
@@ -53,79 +53,108 @@ pub fn listenAndServe(
     comptime handler: RequestHandler(@TypeOf(context)),
 ) !void {
-    try (Server.init()).run(gpa, address, context, handler);
+    try (Server(@TypeOf(context)).init(gpa, address, context, handler)).run();
 }
 
-pub const Server = struct {
-    should_quit: atomic.Atomic(bool),
+pub fn Server(comptime Context: type, comptime handler: RequestHandler(Context)) type {
+    return struct {
+        const Self = @This();
 
-    /// Initializes a new `Server` instance
-    pub fn init() Server {
-        return .{ .should_quit = atomic.Atomic(bool).init(false) };
-    }
+        const Client = ClientFn(Context, handler);
+        should_quit: atomic.Atomic(bool),
 
-    /// Starts listening to new connections and serves the responses
-    /// Cleans up any resources that were allocated during the connection
-    pub fn run(
-        self: *Server,
-        /// Memory allocator, for general usage.
-        /// Will be used to setup an arena to free any request/response data.
         gpa: *Allocator,
-        /// Address the server is listening at
         address: net.Address,
-        /// Runtime context allowing to pass data between handlers.
-        /// Thread-safety must be guaranteed by the caller.
-        context: anytype,
-        /// User defined `Request`/`Response` handler
-        comptime handler: RequestHandler(@TypeOf(context)),
-    ) !void {
-        var stream = net.StreamServer.init(.{ .reuse_address = true });
-        defer stream.deinit();
+        context: Context,
+
+        stream: net.StreamServer,
 
         // client queue to clean up clients after connection is broken/finished
-        const Client = ClientFn(@TypeOf(context), handler);
-        var clients = Queue(*Client).init();
-
-        // Force clean up any remaining clients that are still connected
-        // if an error occured
-        defer while (clients.get()) |node| {
-            const data = node.data;
-            data.stream.close();
-            gpa.destroy(data);
-        };
+        clients: Queue(*Client),
+
+        /// Initializes a new `Server` instance
+        pub fn init(
+            /// Memory allocator, for general usage.
+            /// Will be used to setup an arena to free any request/response data.
+            gpa: *Allocator,
+            /// Address the server is listening at
+            address: net.Address,
+            /// Runtime context allowing to pass data between handlers.
+            /// Thread-safety must be guaranteed by the caller.
+            context: Context,
+        ) Self {
+            return Self{
+                .should_quit = atomic.Atomic(bool).init(false),
+                .gpa = gpa,
+                .address = address,
+                .context = context,
+                .stream = net.StreamServer.init(.{ .reuse_address = true }),
+                .clients = Queue(*Client).init(),
+            };
+        }
 
-        try stream.listen(address);
+        pub fn deinit(self: *Self) void {
```

Comment:
Missing comment on public-facing function. It should mention it will also shutdown the server and close all connections.